cas-client-core/src/main/java/org/jasig/cas/client/util/ThreadLocalXPathExpression.java

```diff
@@ -0,0 +1,87 @@
+package org.jasig.cas.client.util;
+
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+import javax.xml.namespace.NamespaceContext;
+import javax.xml.namespace.QName;
+import javax.xml.xpath.*;
+
+/**
+ * Thread local XPath expression.
+ *
+ * @author Marvin S. Addison
+ * @since 3.3.4
+ */
+public class ThreadLocalXPathExpression extends ThreadLocal<XPathExpression> implements XPathExpression {
```

Comment:
The variables you mentioned should be final; will fix. You are correct on the expensive operation.